python_coreml_stable_diffusion/activation_quantization.py

```diff
@@ -0,0 +1,491 @@
+        quantized_unet.to('cpu')
+        sample_unet_input = {
+            "sample": dataloader[0][0].to('cpu'),
+            "timestep": dataloader[0][1].to('cpu'),
+            "encoder_hidden_states": dataloader[0][2].to('cpu'),
+        }
+
+        logger.info("JIT tracing quantized model")
+        traced_model = torch.jit.trace(quantized_unet, example_inputs=list(sample_unet_input.values()))
+
+        logger.info("Converting to CoreML")
+        coreml_sample_unet_input = {
+            k: v.numpy().astype(np.float16)
+            for k, v in sample_unet_input.items()
+        }
+        coreml_model = convert_to_coreml(traced_model, coreml_sample_unet_input)
+        coreml_filename = f"Stable_Diffusion_version_{args.model_version.replace('/', '_')}_unet.mlpackage"
+        coreml_model.save(os.path.join(args.o, coreml_filename))
+
+        del q_pipe
+
+    del ref_pipe
+    gc.collect()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-o",
+        required=True,
+        help="Output directory to save calibration data and quantization artifacts"
+    )
+    parser.add_argument(
+        "--model-version",
+        required=True,
+        help=
+        ("The pre-trained model checkpoint and configuration to restore. "
+         "For available versions: https://huggingface.co/models?search=stable-diffusion."
+    ))
+    parser.add_argument(
+        "--generate-calibration-data",
+        action="store_true",
+        help="Generate calibration data for UNet model"
+    )
+    parser.add_argument(
+        "--layerwise-sensitivity",
+        action="store_true",
+        help="Compute compression sensitivity per-layer, by quantizing one layer at a time"
+    )
+    parser.add_argument(
+        "--quantize-pytorch",
+        action="store_true",
+        help="Generate activation quantized UNet model by quantizing layers above specified PSNR threshold"
+    )
+    parser.add_argument(
+        "--calibration-nsamples",
+        type=int,
+        help="Number of samples to use for calibrating UNet model"
+    )
+    parser.add_argument("--seed",
+                        "-s",
+                        default=50,
+                        type=int,
+                        help="Random seed to be able to reproduce results"
+    )
+    parser.add_argument("--conv-psnr",
+                        default=40.0,
+                        type=float,
+                        help="PSNR threshold for convolutional layers (default for stabilityai/stable-diffusion-2-1-base)"
+    )
+    parser.add_argument("--attn-psnr",
+                        default=30.0,
+                        type=float,
+                        help="PSNR threshold for attention (Einsum) layers (default for stabilityai/stable-diffusion-2-1-base)"
+    )
+    args = parser.parse_args()
+    if 'xl' in args.model_version:
+        raise NotImplementedError("Not supported for Stable Diffusion XL models")
```

Comment:
Have you tested it for all non xl models? Would it make more sense to raise error for any version that has not been tested. (there are so many versions here: https://huggingface.co/models?search=stable-diffusion)